mmeval/metrics/coco_panoptic.py

```diff
@@ -0,0 +1,656 @@
+                # Directly compute the results and put into `self._results`
+                pq_results = self._compute_single_pq_stats(
+                    prediction, groundtruth)
+                self._results.append(pq_results)
+            else:
+                self._results.append((prediction, groundtruth))
+
+    def compute_metric(self, results: list) -> dict:
+        """Compute the CocoPanoptic metric.
+
+        Args:
+            results (List[tuple] | List[PQStat]): If self.direct_compute is
+                True, it is a list of PQStat, else a list of tuple, each
+                tuple is the prediction and ground truth of an image.
+                The list has already been synced across all ranks.
+
+        Returns:
+            dict: The computed Panoptic Quality metric.
+        """
+        eval_results: OrderedDict = OrderedDict()
+
+        if self.format_only:
+            self.logger.info(f'Results are saved in {osp.dirname(self.outfile_prefix)}')  # type: ignore # yapf: disable # noqa: E501
+            return eval_results
+
+        if self.classes is None:
+            self._get_classes()
+
+        if self.direct_compute:
+            pq_stat_results = results
+        else:
+            pq_stat_results = self._compute_multi_pq_stats(
+                results)  # type: ignore
+
+        # aggregate the results generated in process
+        pq_stat = PQStat()
+        for result in pq_stat_results:
+            pq_stat += result
+
+        metrics = [('All', None), ('Things', True), ('Stuff', False)]
+        pq_results = dict()
+        for name, isthing in metrics:
+            pq_results[name], classwise_results = pq_stat.pq_average(
+                self.categories, isthing=isthing)
+            if name == 'All' and self.classwise:
+                pq_results['classwise'] = classwise_results
+
+        # print tables
+        self._print_panoptic_table(pq_results)
+
+        # process results
+        eval_results['PQ'] = pq_results['All']['pq']
+        eval_results['SQ'] = pq_results['All']['sq']
+        eval_results['RQ'] = pq_results['All']['rq']
+        eval_results['PQ_th'] = pq_results['Things']['pq']
+        eval_results['SQ_th'] = pq_results['Things']['sq']
+        eval_results['RQ_th'] = pq_results['Things']['rq']
+        eval_results['PQ_st'] = pq_results['Stuff']['pq']
+        eval_results['SQ_st'] = pq_results['Stuff']['sq']
+        eval_results['RQ_st'] = pq_results['Stuff']['rq']
+        classwise_results = pq_results.get('classwise')
+        if classwise_results is not None:
+            for k, v in zip(self.classes, classwise_results.values()):  # type: ignore # yapf: disable # noqa: E501
+                eval_results[f'{k}_PQ'] = v['pq']
+
+        # remove result files to save storage space and make dir again
+        # to avoid potential error
+        if not self.keep_results:
+            if osp.exists(self.outfile_prefix):  # type: ignore # yapf: disable # noqa: E501
```

Comment:
what happened to the lint?

Reply:
self.outfile_prefix may be None, and `mypy` cannot pass. These error still happen in `self.classes`, self.thing_classes, etc.

Reply:
If we use a property, these lint errot may not happen anymore